observability/akka-http-metrics/src/test/suite/scala/com/daml/metrics/akkahttp/WebSocketMetricsSpec.scala

```diff
@@ -0,0 +1,342 @@
+  private val drainStreamedMessages: Sink[Message, Future[List[List[ByteString]]]] =
+    Flow[Message]
+      .flatMapConcat {
+        case TextMessage.Strict(text) =>
+          Source.single(List(ByteString(text)))
+        case TextMessage.Streamed(textStream) =>
+          textStream.fold(List[ByteString]())((acc, c) => acc :+ ByteString(c))
+        case BinaryMessage.Strict(data) =>
+          Source.single(List(data))
+        case BinaryMessage.Streamed(dataStream) =>
+          // dataStream.map(_ => 1)
+          dataStream.fold(List[ByteString]())((acc, c) => acc :+ c)
+      }
+      .toMat(Sink.fold(List[List[ByteString]]())((acc, s) => acc :+ s))((_, mat2) => mat2)
+
+  private val getMessagesReceivedTotalValue: TestMetrics => Long = (metrics: TestMetrics) =>
+    metrics.messagesReceivedTotalValue
+
+  private val getMessagesSentTotalValue: TestMetrics => Long = (metrics: TestMetrics) =>
+    metrics.messagesSentTotalValue
+
+  private val getMessagesReceivedBytesTotalValue: TestMetrics => Long =
+    (metrics: TestMetrics) => metrics.messagesReceivedBytesTotalValue
+
+  private val getMessagesSentBytesTotalValue: TestMetrics => Long = (metrics: TestMetrics) =>
+    metrics.messagesSentBytesTotalValue
+
+  private def checkCountThroughFlow(
+      messages: List[Message],
+      metricExtractor: (TestMetrics) => Long,
+      expected: Long,
+  ): Future[Assertion] = {
+    withDuplicatingFlowAndMetrics { (flow, metrics) =>
+      val done = Source(messages).via(flow).runWith(drainStreamedMessages)
+      done.map { _ =>
+        metricExtractor(metrics) should be(expected)
+      }
+    }
+  }
+
+  private def checkCountThroughFlow(
+      message: Message,
+      metricExtractor: TestMetrics => Long,
+      expected: Long,
+  ): Future[Assertion] = {
+    checkCountThroughFlow(List(message), metricExtractor, expected)
+  }
+
+  "websocket_message_received_total" should {
+    "count passing strict text messages" in {
+      checkCountThroughFlow(strictTextMessage, getMessagesReceivedTotalValue, 1)
+    }
+
+    "count passing streamed text messages" in {
+      checkCountThroughFlow(streamedTextMessage, getMessagesReceivedTotalValue, 1)
+    }
+
+    "count passing strict binary messages" in {
+      checkCountThroughFlow(strictBinaryMessage, getMessagesReceivedTotalValue, 1)
+    }
+
+    "count passing streamed binary messages" in {
+      checkCountThroughFlow(streamedBinaryMessage, getMessagesReceivedTotalValue, 1)
+    }
+
+    "count a set of passing messages" in {
+      checkCountThroughFlow(
+        List[Message](
+          strictTextMessage,
+          streamedTextMessage,
+          strictBinaryMessage,
+          streamedBinaryMessage,
+        ),
+        getMessagesReceivedTotalValue,
+        4,
+      )
+    }
+  }
+
+  "websocket_message_sent_total" should {
+    "count passing strict text messages" in {
+      checkCountThroughFlow(strictTextMessage, getMessagesSentTotalValue, 2)
```

Comment:
Why is the expected value always 2 here?

Reply:
The test flow duplicate the messages.
It is mainly done to avoid copy-paste errors: the result for the received metrics are different than the sent metrics.

I'll try to make it more visible.